### ISSUE TITLE
Yesod devel doesn't respect default extensions specified in Cabal file

### ISSUE THREAD

**Issue by mrkkrp** [yesodweb/yesod#1107]
I've specified some `default-extensions` in Cabal file of my project. The project compiles fine but `stack exec -- yesod devel` doesn't work (reports lexical error as if `CPP` were not enabled). Is this Cabal feature not supported?

**Comment by snoyberg** [comment]
I haven't seen this behavior previously, and in fact the default scaffolding using CPP and TemplateHaskell and works with `yesod devel`. Can you provide a reproducing case?

**Comment by mrkkrp** [comment]
Several things I have to clarify. Cabal supports `default-extensions` beginning from version 1.10. I can't find docs that state this, but Cabal itself prints this:

```
Warning: For packages using 'cabal-version: >= 1.10' the 'extensions' field is
deprecated. The new 'default-extensions' field lists extensions that are used
in all modules in the component, while the 'other-extensions' field lists
extensions that are used in some modules, e.g. via the {-# LANGUAGE #-}
pragma.
```

You won't see this if you are specifying `cabal-version: >= 1.8` (current scaffolding practice). But the fact is, `default-extensions` seems to be preferable way to enable extensions everywhere now and it would be great if `yesod-devel` supported it.

**Comment by snoyberg** [comment]
This just muddies the waters more, a reproducing case is necessary. Are you finding that something is buildable with cabal but not with yesod devel? That's quite strange, since under the surface yesod devel is just using cabal for building.

**Comment by mrkkrp** [comment]
So, to reproduce this it's enough to specify `cabal-version: >= 1.10` and replace `extensions` → `default-extensions` as recommended. Then run

```
$ stack exec -- yesod devel
```

**Comment by mrkkrp** [comment]
I tried to reproduce this beginning from `yesod-postgres` template but I couldn't. Here is however project I'm working on and there it doesn't work (branch `default-exntension-malfunction-demo`):

https://github.com/svod-music/svod/tree/default-extension-malfunction-demo

Somehow it doesn't work there. It builds with `stack build` fine, but when I run `stack exec -- yesod devel`:

```
Yesod devel server. Type 'quit' to quit
Application can be accessed at:

http://127.0.0.1:3000
https://127.0.0.1:3443
If you wish to test https capabilities, you should set the following variable:
  export APPROOT=https://127.0.0.1:3443

Warning: The package list for 'hackage.haskell.org' is 36.2 days old.
Run 'cabal update' to get the latest list of available packages.
Resolving dependencies...
Configuring svod-0.1.0...
Rebuilding application... (using cabal)
Starting development server...

Foundation.hs:52:1:
    Parse error: naked expression at top level
    Perhaps you intended to use TemplateHaskell
Exit code: ExitFailure 1
quit
```

I feel a bit perplexed right now, many things don't work and I'm tired. I'm sorry that I cannot give you clear list of steps how to reproduce this beginning from fresh scaffolding using `yesod-postgres` template (that would be more useful for you, I understand). But that site is essentially edited (not heavily) scaffolding, you can just take a look at that. I'll return to this issue and I'll try to find out why exactly it doesn't work.

**Comment by snoyberg** [comment]
I don't see CPP in the default-extensions list in your cabal file.

**Comment by mrkkrp** [comment]
Well this demonstrates just the case with `TemplateHaskell`, I could similarly remove `CPP` from individual files (I currently keep pramas in files) where it's used and add it to `default-extensions`, then it would fail with different message (when actual preprocessor macros would be encountered, for example `#if`). What this setup should demonstrate it that it has `TemplateHaskell` in `default-extensions` and it builds fine, but `yesod devel` fails to recompile the project when it encounters `mkYesodData` in `Foundation.hs`— if `TemplateHaskell` was enabled it wouldn't fail. I think it demonstrates the issue.

**Comment by mrkkrp** [comment]
I still don't see why this happens, but maybe the fact that error message is printed *after* “Starting development server...” phrase may have any significance? “Rebuilding application... (using cabal)” obviously succeeds as does `stack build`. Why `yesod devel` rebuilds `Foundation.hs` after starting development server? Does it automatically consider it modified (despite the fact that it *has just rebuilt it using Cabal*)? I'm not sure why this rebuild happens at all.

**Comment by snoyberg** [comment]
In order to look into this, I'll need a much more minimal case, ideally just a few well-documented tweaks to the scaffolding.

**Comment by mrkkrp** [comment]
I've found out what causes this issue. In `app/devel.hs` I removed `PackageImports` pragma and name of package from import. This somehow makes this difference. It would be great if you added a comment explaining why this is needed.

All in all, this is certainly my fault (I like to tweak and change everything and sometimes I get into such situations, this is also why I like Haskell, it can tell me immediately when I do something wrong, well, almost always). Sorry for annoyance! Closing.
